src/mono/mono/mini/interp/interp-intrins.h

```diff
@@ -6,6 +6,123 @@
 
 #include "interp-internals.h"
 
+#ifdef __GNUC__
+static inline gint32
+interp_intrins_clz (guint32 val)
+{
+	if (val == 0)
+		return 32;
+        return __builtin_clz (val);
+}
+
+static inline gint32
+interp_intrins_clzl (guint64 val)
+{
+	if (val == 0)
+		return 64;
+        return __builtin_clzl (val);
+}
+
+static inline gint32
+interp_intrins_ctz (guint32 val)
+{
+	if (val == 0)
+		return 32;
+        return __builtin_ctz (val);
+}
+
+static inline gint32
+interp_intrins_ctzl (guint64 val)
+{
+	if (val == 0)
+		return 64;
+        return __builtin_ctzl (val);
+}
+
+static inline gint32
+interp_intrins_popcount (guint32 val)
+{
+        return __builtin_popcount (val);
+}
+
+static inline gint32
+interp_intrins_popcountl (guint64 val)
+{
+        return __builtin_popcountl (val);
+}
+#else
+static inline gint32
+interp_intrins_clz (guint32 val)
+{
+	gint32 count = 0;
+        while (val) {
+                val = val >> 1;
+                count++;
+        }
+        return 32 - count;
+}
+
+static inline gint32
+interp_intrins_clzl (guint64 val)
+{
+	gint32 count = 0;
+        while (val) {
+                val = val >> 1;
+                count++;
+        }
+        return 64 - count;
+}
+
+static inline gint32
+interp_intrins_ctz (guint32 val)
+{
+	if (val == 0)
+		return 32;
+	gint32 count = 0;
+	while ((val & 1) == 0) {
+		val = val >> 1;
+		count++;
+	}
+	return count;
+}
+
+static inline gint32
+interp_intrins_ctzl (guint64 val)
+{
+	if (val == 0)
+		return 64;
+	gint32 count = 0;
+	while ((val & 1) == 0) {
+		val = val >> 1;
+		count++;
+	}
+	return count;
+}
+
+static inline gint32
+interp_intrins_popcount (guint32 val)
+{
+	gint32 count = 0;
+	while (val != 0) {
+		count += val & 1;
+		val = val >> 1;
+	}
+	return count;
+}
+
+static inline gint32
+interp_intrins_popcountl (guint64 val)
+{
+	gint32 count = 0;
+	while (val != 0) {
+		count += val & 1;
+		val = val >> 1;
+	}
+	return count;
+}
```

Comment:
nit: Most of these have far more efficient fallbacks.

The `BitOperations::*` helpers defined starting at the following in RyuJIT represent the efficient fallbacks: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/utils.cpp#L2538

The split is `MSVC` vs `clang/gcc` with https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/utils.h#L743 covering the full suite of APIs we special case today (notably you're missing `Log2`).

Reply:
I agree. However we don't have a strong use case for interp on windows for now, we don't even test it on CI. So I'd rather leave it as a TODO for now, together with the v128 intrinsics which are also `gnuc` only.

Reply:
`Log2` was implemented quite efficiently based on `Lzcnt` so I decided to avoid adding another intrinsic.

Reply:
> However we don't have a strong use case for interp on windows for now

👍, makes sense

> Log2 was implemented quite efficiently based on Lzcnt so I decided to avoid adding another intrinsic

Right. The general case, however, is that needing to process the IL, let alone optimize it, is "significantly more expensive" than recognizing the API directly.

That is, `Log2` is efficiently compiled down. However, its implementation is quite a lot of IL and the JIT/interpreter has to do non-trivial work to get there.

The current C# implementation is:
```csharp
// The 0->0 contract is fulfilled by setting the LSB to 1.
// Log(1) is 0, and setting the LSB for values > 1 does not change the log2 result.
value |= 1;

// value    lzcnt   actual  expected
// ..0001   31      31-31    0
// ..0010   30      31-30    1
// 0010..    2      31-2    29
// 0100..    1      31-1    30
// 1000..    0      31-0    31
if (Lzcnt.IsSupported)
{
    return 31 ^ (int)Lzcnt.LeadingZeroCount(value);
}

if (ArmBase.IsSupported)
{
    return 31 ^ ArmBase.LeadingZeroCount(value);
}

if (WasmBase.IsSupported)
{
    return 31 ^ WasmBase.LeadingZeroCount(value);
}

// BSR returns the log2 result directly. However BSR is slower than LZCNT
// on AMD processors, so we leave it as a fallback only.
if (X86Base.IsSupported)
{
    return (int)X86Base.BitScanReverse(value);
}

// Fallback contract is 0->0
return Log2SoftwareFallback(value);
```

Even with trimming, we're still left with several "dead" branches and other things that have to be handled:
```csharp
value |= 1U;
if (Lzcnt.IsSupported)
  return 31 ^ (int) Lzcnt.LeadingZeroCount(value);
if (true) { } // ArmBase.IsSupported
if (true) { } // WasmBase.IsSupported
return X86Base.IsSupported ? (int) X86Base.BitScanReverse(value) : BitOperations.Log2SoftwareFallback(value);
```

So, we still have to process the `or`, the `Lzcnt.IsSupported` check, then the `31 ^ Lzcnt`. On Wasm or Arm64, this means we also have to process and eliminate 1-2 dead blocks. If `Lzcnt` isn't supported, we also have to process another branch, etc.

All of this before we can do any simple constant folding or even get to the efficient "lzcnt" intrinsic.

-- https://github.com/dotnet/runtime/issues/82822 tracks the linker leaving the "dead blocks" after substitution.
-- There looks to be an issue where `X86Base.IsSupported` is not being substituted, even though it could be. Will see about fixing that.

Reply:
Intrinsified also Log2. The Log2 is also 60% faster this way on interpreter.